### PR TITLE
Described data copy cases

### PR DESCRIPTION
**Pull request by olegkkruglov**
## Description

This PR adds description for cases where data are implicitly copied to documentation.
It solves issue https://github.com/uxlfoundation/scikit-learn-intelex/issues/4

<details>
<summary>Checklist:</summary>

**Completeness and readability**

- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes or created a separate PR with updates and provided its number in the description, if necessary.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have resolved any merge conflicts that might occur with the base branch.

</details>

### CODE REVIEW

File: doc/sources/input-types.rst [hunk 1]
```diff
@@ -39,3 +39,13 @@ like `Polars <https://pola.rs>`__.
 Extension currently does not offer accelerated routines for input types not listed
 here - when receiving an unsupported class, estimators will fall back to stock Scikit-Learn to
 handle it, so make sure to convert them to a supported type when using Extension.
+
+.. warning::
+  In certain cases data could be copied during transition to oneDAL tables.
```
Comment (david-cortes-intel) [inline doc/sources/input-types.rst]: I think this might not sound very descriptive to an external user reading the documentation. OneDAL tables are an internal implementation detail -  they aren't interfaced directly when using sklearnex.

Perhaps it could be reworded along the lines of "data passed to estimators might be copied/duplicated during calls to methods such as fit/predict under some circumstances".

File: doc/sources/input-types.rst [hunk 2]
```diff
@@ -39,3 +39,13 @@ like `Polars <https://pola.rs>`__.
...
+  It would affect performance. The affected cases are listed below.
+
+  - Non-contiguous NumPy array (C or F)
```
Comment (david-cortes-intel) [inline doc/sources/input-types.rst]:
```suggestion
  - Non-contiguous NumPy array - i.e. where strides are wider than one element across both rows and columns
```

File: doc/sources/input-types.rst [hunk 3]
```diff
@@ -39,3 +39,13 @@ like `Polars <https://pola.rs>`__.
...
+  - SciPy CSR matrix / CSR array
```
Comment (david-cortes-intel) [inline doc/sources/input-types.rst]: Here it could mention that all sparse matrices are copied, but CSR ones copy only the indices.

File: doc/sources/input-types.rst [hunk 4]
```diff
@@ -39,3 +39,13 @@ like `Polars <https://pola.rs>`__.
...
+  - Heterogeneous NumPy array
+  - If :ref:`Array API <array_api>` is not enabled then data from GPU devices are always copied to the host device and then result table
```
Comment (david-cortes-intel) [inline doc/sources/input-types.rst:51]: This is not actually the case - for example, passing DPNP arrays that are on GPU to an estimator with GPU support results in the data being used on-device without copying.

File: doc/sources/input-types.rst [hunk 5]
```diff
@@ -39,3 +39,13 @@ like `Polars <https://pola.rs>`__.
...
 here - when receiving an unsupported class, estimators will fall back to stock Scikit-Learn to
 handle it, so make sure to convert them to a supported type when using Extension.
+
+.. warning::
+  In certain cases data could be copied during transition to oneDAL tables.
+  It would affect performance. The affected cases are listed below.
+
+  - Non-contiguous NumPy array (C or F)
+  - SciPy CSR matrix / CSR array
+  - Heterogeneous NumPy array
```
Comment (david-cortes-intel) [inline doc/sources/input-types.rst:49]: Would masked arrays also get copied? If so, perhaps could add it as another entry.

File: doc/sources/input-types.rst [hunk 6]
```diff
@@ -39,3 +39,13 @@ like `Polars <https://pola.rs>`__.
...
+  - If :ref:`Array API <array_api>` is not enabled then data from GPU devices are always copied to the host device and then result table 
+    (for applicable methods) is copied to the source device.
```
Comment (icfaust) [inline doc/sources/input-types.rst:52]: i agree with @david-cortes-intel , if we want to keep this version, we also are missing fp64 -> fp32 for low precision devices

File: doc/sources/input-types.rst (hunk 4 above)
Comment (olegkkruglov) [reply doc/sources/input-types.rst:51]: I thought that this `_transfer_to_host` occurs here always if `onedal_array_api == False`. Could you please correct if I'm missing something?

https://github.com/uxlfoundation/scikit-learn-intelex/blob/fada56a2a44351c49e1034fecad083ae69af39e9/sklearnex/_device_offload.py#L138

Comment (david-cortes-intel) [reply doc/sources/input-types.rst:51]: @icfaust It does look like @olegkkruglov 's comment is right: if passing something like a DPNP array on GPU, there is a data transfer from GPU to host here:
https://github.com/uxlfoundation/scikit-learn-intelex/blob/2874cb1fa43bf0090f32a8b53295ae04ecb22bcf/sklearnex/_device_offload.py#L140
.. even though the log is called with `transferred_to_host=False`:
https://github.com/uxlfoundation/scikit-learn-intelex/blob/2874cb1fa43bf0090f32a8b53295ae04ecb22bcf/sklearnex/_device_offload.py#L156

Any comments about that?

File: doc/sources/input-types.rst [hunk 7]
```diff
@@ -39,3 +39,13 @@ like `Polars <https://pola.rs>`__.
 Extension currently does not offer accelerated routines for input types not listed
 here - when receiving an unsupported class, estimators will fall back to stock Scikit-Learn to
 handle it, so make sure to convert them to a supported type when using Extension.
+
+.. warning::
+  In certain cases data passed to estimators might be copied/duplicated during calls to methods such as fit/predict under some circumstances.
```
Comment (david-cortes-intel) [inline doc/sources/input-types.rst]:
```suggestion
  In some cases data passed to estimators might be copied/duplicated during calls to methods such as fit/predict.
```

File: doc/sources/input-types.rst (hunk 4 above)
Comment (icfaust) [reply doc/sources/input-types.rst:51]: There are a couple things at play there:

1. Array API support is not the default, meaning the ```_onedal_cpu_supported``` and ```_oneda_gpu_supported``` may not properly evaluate with the given input. ```_transfer_to_host``` enables for those previously mentioned methods to operate properly regardless of the circumstances.
2. Until 0.18, dpnp lacked array API support. This means that falling back to scikit-learn required special consideration. This should be updated and I'll make a ticket for it. (I may be able to address this in November, but seems unlikely)
3.  This is behavior was introduced in #2113. Give it a look. I sort of understood it's utility came from how the data approaches oneDAL or scikit-learn. In this branch the data is not *hostargs **hostkwargs, but *args and **kwargs.

Comment (david-cortes-intel) [reply doc/sources/input-types.rst:51]: @icfaust Sorry, I linked the wrong line.

What I meant was this other line here:
https://github.com/uxlfoundation/scikit-learn-intelex/blob/2874cb1fa43bf0090f32a8b53295ae04ecb22bcf/sklearnex/_device_offload.py#L151

If passing a GPU DPNP array, by the moment it reaches that part:
* The GPU data would be transferred to CPU as a NumPy array already by this point: https://github.com/uxlfoundation/scikit-learn-intelex/blob/2874cb1fa43bf0090f32a8b53295ae04ecb22bcf/sklearnex/_device_offload.py#L140
* The log will nevertheless be called with `transferred_to_host=False`.
* The data will be passed to sklearnex directly, since it has been established that the op will be supported.
* Next, the data will be transferred again from CPU to GPU, and the compute will happen on GPU.

Comment (icfaust) [reply doc/sources/input-types.rst:51]: Looks a like a bug, good catch. I look forward to your PR.

File: doc/sources/input-types.rst (hunk 5 above)
Comment (olegkkruglov) [reply doc/sources/input-types.rst:49]: If you mean numpy masked arrays then they are handled as numpy arrays here
https://github.com/uxlfoundation/scikit-learn-intelex/blob/cda899f4500da545090143c602714eb71dc72962/onedal/datatypes/numpy/data_conversion.cpp#L168
so, no special cases are necessary. If you mean something else, could you please clarify?

Comment (david-cortes-intel) [reply doc/sources/input-types.rst:49]: Thanks. Looks like it ends up simply ignoring the mask. OTOH looks like sklearn does the same for these classes, so nothing to worry about for now.

**Review by david-cortes-intel — APPROVED** [review]
